src/handshake-xx-fallback.ts

```diff
@@ -0,0 +1,89 @@
+import { Buffer } from "buffer";
+
+import { Handshake as XXHandshake } from "./handshake-xx";
+import { XX } from "./handshakes/xx";
+import { KeyPair, PeerId } from "./@types/libp2p";
+import { bytes, bytes32 } from "./@types/basic";
+import {
+  createHandshakePayload,
+  getHandshakePayload,
+  signEarlyDataPayload,
+  signPayload,
+  verifySignedPayload,
+} from "./utils";
+import { logger } from "./logger";
+import { WrappedConnection } from "./noise";
+import {decode0, decode1, encode1} from "./encoder";
+
+export class Handshake extends XXHandshake {
+  private ephemeralKeys: KeyPair;
+  private initialMsg: bytes;
+
+  constructor(
+    isInitiator: boolean,
+    libp2pPrivateKey: bytes,
+    libp2pPublicKey: bytes,
+    prologue: bytes32,
+    staticKeys: KeyPair,
+    connection: WrappedConnection,
+    remotePeer: PeerId,
+    ephemeralKeys: KeyPair,
+    initialMsg: bytes,
+    handshake?: XX,
+  ) {
+    super(isInitiator, libp2pPrivateKey, libp2pPublicKey, prologue, staticKeys, connection, remotePeer, handshake);
+    this.ephemeralKeys = ephemeralKeys;
+    this.initialMsg = initialMsg;
+  }
+
+  // stage 0
+  public async propose(): Promise<void> {
+    if (this.isInitiator) {
+      this.xx.sendMessage(this.session, Buffer.alloc(0), this.ephemeralKeys);
+      logger("XX Fallback Stage 0 - Initialized state as the first message was sent by initiator.");
+    } else {
+      logger("XX Fallback Stage 0 - Responder waiting to receive first message...");
+      const receivedMessageBuffer = decode0(this.initialMsg);
+      this.xx.recvMessage(this.session, {
+        ne: receivedMessageBuffer.ne,
+        ns: Buffer.alloc(0),
+        ciphertext: Buffer.alloc(0),
+      });
+      logger("XX Fallback Stage 0 - Responder received first message.");
+    }
+  }
+
+  // stage 1
+  public async exchange(): Promise<void> {
+    if (this.isInitiator) {
+      logger('XX Fallback Stage 1 - Initiator waiting to receive first message from responder...');
+      const receivedMessageBuffer = decode1((await this.connection.readLP()).slice());
```

Comment:
why slice? It will create new buffer instead of using existing (I don't think connection will care if you modify that)